AppKit/CPWindow/CPWindow.j

```diff
@@ -3844,6 +3819,393 @@ var interpolate = function(fromValue, toValue, progress)
 
 @end
 
+@implementation CPWindow (TrackingAreaAdditions)
+
+- (void)_addTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _addTrackingArea:trackingAreas[i]];
+}
+
+- (void)_removeTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _removeTrackingArea:trackingAreas[i]];
+}
+
+- (void)_addTrackingArea:(CPTrackingArea)trackingArea
+{
+    var trackingAreaView = [trackingArea view];
+    
+    if (![_trackingAreaViews containsObject:trackingAreaView])
+        [_trackingAreaViews addObject:trackingAreaView];
+    
+    // If CPTrackingAssumeInside option is set, put the tracking area in the _mouseEnteredStack
+    
+    if ([trackingArea options] & CPTrackingAssumeInside)
+        [_mouseEnteredStack addObject:trackingArea];
+}
+
+- (void)_removeTrackingArea:(CPTrackingArea)trackingArea
+{
+    // If mouse is in the tracking area, we remove it from the stack to avoid to fire a future mouseExited event
+    
+    [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
+    
+    var trackingAreaView = [trackingArea view];
+    
+    [_trackingAreaViews removeObjectIdenticalTo:trackingAreaView];
+}
+
+- (void)_handleTrackingAreaEvent:(CPEvent)anEvent
```

Comment:
In general, readability and maintainability should always take precedence over saving a millisecond of execution time. And studies have shown that we humans are extremely bad at determining where performance bottlenecks are. We tend to micro-optimize things that don't matter, and miss the real bottlenecks.

Case in point: in terms of method call performance hits, what's killing us is using `containsObject` with arrays. First of all, it should be `containsIdenticalObject`, because `containsObject` compares using `isEqual` when `===` is native and sufficient. Second, every call to `containsObject` goes through 3 method calls before it even starts comparing objects, and then there are 2 more method calls for each item in the array.

So if you're worried about performance, don't worry about 2 extra method calls per event, worry more about using `containsObject`. There really is not point in using it, you should just use the native code `anArray.indexOf(object) >= 0` instead, which will probably be 10x faster.